File: server/services/transactionService.js
const Account = require('../models/account');
const Transaction = require('../models/transaction');
const User = require('../models/user');

// Exchange rate: 1 gold bar = $7000
const GOLD_BAR_VALUE = 7000;

const TransactionService = {
  // Transfer money between checking and savings accounts
  transfer(fromAccountId, toAccountId, amount, description = 'Transfer') {
    if (amount <= 0) {
      throw new Error('Amount must be positive');
    }

    const fromAccount = Account.findById(fromAccountId);
    const toAccount = Account.findById(toAccountId);

    if (!fromAccount) {
      throw new Error(`Source account not found: ${fromAccountId}`);
    }
    if (!toAccount) {
      throw new Error(`Destination account not found: ${toAccountId}`);
    }

    // Cannot transfer to/from treasure chest (it's gold bars only)
    if (fromAccount.type === 'treasure_chest' || toAccount.type === 'treasure_chest') {
      throw new Error('Cannot transfer to/from treasure chest. Use exchangeGold instead.');
    }

    // Check sufficient balance
    if (fromAccount.balance < amount) {
      throw new Error('Insufficient balance');
    }

    // Perform transfer
    Account.subtractFromBalance(fromAccountId, amount);
    Account.addToBalance(toAccountId, amount);

    // Record transaction
    return Transaction.create(fromAccountId, toAccountId, amount, 'transfer', description);
  },

  // Receive external payment (deposit)
  deposit(toAccountId, amount, description = 'Deposit') {
    if (amount <= 0) {
      throw new Error('Amount must be positive');
    }

    const toAccount = Account.findById(toAccountId);
    if (!toAccount) {
      throw new Error(`Account not found: ${toAccountId}`);
    }

    // Cannot deposit to treasure chest directly
    if (toAccount.type === 'treasure_chest') {
      throw new Error('Cannot deposit money to treasure chest');
    }

    // Add to balance
    Account.addToBalance(toAccountId, amount);

    // Record transaction (null from_account means external)
    return Transaction.create(null, toAccountId, amount, 'deposit', description);
  },

  // Make external payment (withdrawal)
  withdraw(fromAccountId, amount, description = 'Payment') {
    if (amount <= 0) {
      throw new Error('Amount must be positive');
    }

    const fromAccount = Account.findById(fromAccountId);
    if (!fromAccount) {
      throw new Error(`Account not found: ${fromAccountId}`);
    }

    // Cannot withdraw from treasure chest directly
    if (fromAccount.type === 'treasure_chest') {
      throw new Error('Cannot withdraw from treasure chest. Use exchangeGold first.');
    }

    // Check sufficient balance
    if (fromAccount.balance < amount) {
      throw new Error('Insufficient balance');
    }

    // Subtract from balance
    Account.subtractFromBalance(fromAccountId, amount);

    // Record transaction (null to_account means external)
    return Transaction.create(fromAccountId, null, amount, 'withdrawal', description);
  },

  // Collect a gold bar (from treasure chest in game)
  collectGoldBar(userId) {
    if (!User.exists(userId)) {
      throw new Error(`User not found: ${userId}`);
    }

    const treasureChest = Account.findByUserIdAndType(userId, 'treasure_chest');
    if (!treasureChest) {
      throw new Error('Treasure chest account not found');
    }

    // Add 1 gold bar
    Account.addToBalance(treasureChest.id, 1);

    // Record transaction
    return Transaction.create(null, treasureChest.id, 1, 'deposit', 'Gold bar collected from treasure chest');
  },

  // Exchange gold bars for cash
  exchangeGold(userId, bars, toAccountType = 'checking') {
    if (bars <= 0) {
      throw new Error('Number of bars must be positive');
    }

    if (!Number.isInteger(bars)) {
      throw new Error('Number of bars must be a whole number');
    }

    if (toAccountType !== 'checking' && toAccountType !== 'savings') {
      throw new Error('Can only exchange gold to checking or savings account');
    }

    if (!User.exists(userId)) {
      throw new Error(`User not found: ${userId}`);
    }

    const treasureChest = Account.findByUserIdAndType(userId, 'treasure_chest');
    const targetAccount = Account.findByUserIdAndType(userId, toAccountType);

    if (!treasureChest) {
      throw new Error('Treasure chest account not found');
    }
    if (!targetAccount) {
      throw new Error(`${toAccountType} account not found`);
    }

    // Check sufficient gold bars
    if (treasureChest.balance < bars) {
      throw new Error(`Insufficient gold bars. Have: ${treasureChest.balance}, need: ${bars}`);
    }

    const cashAmount = bars * GOLD_BAR_VALUE;

    // Remove gold bars from treasure chest
    Account.subtractFromBalance(treasureChest.id, bars);

    // Add cash to target account
    Account.addToBalance(targetAccount.id, cashAmount);

    // Record transaction
    return Transaction.create(
      treasureChest.id,
      targetAccount.id,
      cashAmount,
      'gold_exchange',
      `Exchanged ${bars} gold bar(s) for $${cashAmount}`
    );
  },

  // Get transaction history for a user
  getTransactionHistory(userId, limit = 50) {
    if (!User.exists(userId)) {
      throw new Error(`User not found: ${userId}`);
    }
    return Transaction.findByUserId(userId, limit);
  },

  // Get transaction history for a specific account
  getAccountTransactions(accountId, limit = 50) {
    const account = Account.findById(accountId);
    if (!account) {
      throw new Error(`Account not found: ${accountId}`);
    }
    return Transaction.findByAccountId(accountId, limit);
  },

  // Get the gold bar exchange rate
  getGoldBarValue() {
    return GOLD_BAR_VALUE;
  },

  // Send money from one user to another
  sendMoney(fromUserId, toUserId, amount, fromAccountType = 'checking', toAccountType = 'checking', description = '') {
    if (amount <= 0) {
      throw new Error('Amount must be positive');
    }

    if (fromUserId === toUserId) {
      throw new Error('Cannot send money to yourself. Use transfer instead.');
    }

    if (!User.exists(fromUserId)) {
      throw new Error(`Sender not found: ${fromUserId}`);
    }
    if (!User.exists(toUserId)) {
      throw new Error(`Recipient not found: ${toUserId}`);
    }

    const fromAccount = Account.findByUserIdAndType(fromUserId, fromAccountType);
    const toAccount = Account.findByUserIdAndType(toUserId, toAccountType);

    if (!fromAccount) {
      throw new Error(`Sender's ${fromAccountType} account not found`);
    }
    if (!toAccount) {
      throw new Error(`Recipient's ${toAccountType} account not found`);
    }

    if (fromAccountType === 'treasure_chest' || toAccountType === 'treasure_chest') {
      throw new Error('Cannot send money to/from treasure chest');
    }

    if (fromAccount.balance < amount) {
      throw new Error('Insufficient balance');
    }

    // Perform transfer
    Account.subtractFromBalance(fromAccount.id, amount);
    Account.addToBalance(toAccount.id, amount);

    // Record transaction
    const desc = description || `Payment to ${toUserId}`;
    return Transaction.create(fromAccount.id, toAccount.id, amount, 'transfer', desc);
  },

  // Find users (for recipient lookup)
  findUsers(searchTerm) {
    const allUsers = User.findAll();
    if (!searchTerm) {
      return allUsers;
    }
    const term = searchTerm.toLowerCase();
    return allUsers.filter(u =>
      u.id.toLowerCase().includes(term) ||
      u.name.toLowerCase().includes(term)
    );
  }
};

module.exports = TransactionService;

File: server/services/transaction_service.py
from sqlalchemy.orm import Session
from fastapi import HTTPException
from models.transaction import Transaction
from models.account import Account
from models.user import User
from .account_service import AccountService


# Gold exchange rate: 1 gold bar = $7000
GOLD_BAR_VALUE = 7000


class TransactionService:
    def __init__(self, db: Session):
        self.db = db
        self.account_service = AccountService(db)

    def get_gold_bar_value(self) -> int:
        """Get the value of one gold bar."""
        return GOLD_BAR_VALUE

    def transfer(
        self,
        from_account_id: int,
        to_account_id: int,
        amount: float,
        description: str = None
    ) -> Transaction:
        """Transfer between accounts (not treasure chest)."""
        from_account = self.account_service.get_account(from_account_id)
        to_account = self.account_service.get_account(to_account_id)

        # Validate not treasure chest
        if from_account.type == "treasure_chest" or to_account.type == "treasure_chest":
            raise HTTPException(
                status_code=400,
                detail="Cannot transfer to/from treasure chest. Use gold exchange instead."
            )

        # Validate same user
        if from_account.user_id != to_account.user_id:
            raise HTTPException(
                status_code=400,
                detail="Cannot transfer between different users. Use send money instead."
            )

        if from_account.is_loan:
            raise HTTPException(
                status_code=400,
                detail="Cannot transfer from credit card accounts."
            )

        # Validate sufficient funds
        if from_account.balance < amount:
            raise HTTPException(status_code=400, detail="Insufficient funds")

        # Perform transfer
        from_account.balance -= amount
        if to_account.is_loan:
            to_account.balance = max(to_account.balance - amount, 0)
        else:
            to_account.balance += amount

        # Record transaction
        transaction = Transaction(
            from_account_id=from_account_id,
            to_account_id=to_account_id,
            amount=amount,
            type="transfer",
            description=description or f"Transfer from {from_account.name} to {to_account.name}"
        )
        self.db.add(transaction)
        self.db.commit()
        self.db.refresh(transaction)
        return transaction

    def deposit(
        self,
        account_id: int,
        amount: float,
        description: str = None
    ) -> Transaction:
        """Deposit funds into account (not treasure chest)."""
        account = self.account_service.get_account(account_id)

        if account.type == "treasure_chest":
            raise HTTPException(
                status_code=400,
                detail="Cannot deposit to treasure chest. Use collect gold instead."
            )

        if account.is_loan:
            account.balance = max(account.balance - amount, 0)
        else:
            account.balance += amount

        transaction = Transaction(
            from_account_id=None,
            to_account_id=account_id,
            amount=amount,
            type="deposit",
            description=description or "External deposit"
        )
        self.db.add(transaction)
        self.db.commit()
        self.db.refresh(transaction)
        return transaction

    def withdraw(
        self,
        account_id: int,
        amount: float,
        description: str = None
    ) -> Transaction:
        """Withdraw funds from account (not treasure chest)."""
        account = self.account_service.get_account(account_id)

        if account.type == "treasure_chest":
            raise HTTPException(
                status_code=400,
                detail="Cannot withdraw from treasure chest. Use gold exchange instead."
            )

        if account.is_loan:
            account.balance += amount
        else:
            if account.balance < amount:
                raise HTTPException(status_code=400, detail="Insufficient funds")
            account.balance -= amount

        transaction = Transaction(
            from_account_id=account_id,
            to_account_id=None,
            amount=amount,
            type="withdrawal",
            description=description or "External withdrawal"
        )
        self.db.add(transaction)
        self.db.commit()
        self.db.refresh(transaction)
        return transaction

    def collect_gold_bar(self, user_id: str) -> Transaction:
        """Add one gold bar to user's treasure chest."""
        treasure_chest = self.account_service.get_account_by_type(user_id, "treasure_chest")

        treasure_chest.balance += 1

        transaction = Transaction(
            from_account_id=None,
            to_account_id=treasure_chest.id,
            amount=1,
            type="deposit",
            description="Gold bar collected from game"
        )
        self.db.add(transaction)
        self.db.commit()
        self.db.refresh(transaction)
        return transaction

    def exchange_gold(
        self,
        user_id: str,
        bars: int,
        to_account_type: str
    ) -> Transaction:
        """Exchange gold bars for cash."""
        if to_account_type not in ["checking", "savings"]:
            raise HTTPException(
                status_code=400,
                detail="Can only exchange gold to checking or savings account"
            )

        treasure_chest = self.account_service.get_account_by_type(user_id, "treasure_chest")
        to_account = self.account_service.get_account_by_type(user_id, to_account_type)

        if treasure_chest.balance < bars:
            raise HTTPException(
                status_code=400,
                detail=f"Insufficient gold bars. Have {int(treasure_chest.balance)}, need {bars}"
            )

        cash_amount = bars * GOLD_BAR_VALUE

        treasure_chest.balance -= bars
        to_account.balance += cash_amount

        transaction = Transaction(
            from_account_id=treasure_chest.id,
            to_account_id=to_account.id,
            amount=bars,
            type="gold_exchange",
            description=f"Exchanged {bars} gold bars for ${cash_amount}"
        )
        self.db.add(transaction)
        self.db.commit()
        self.db.refresh(transaction)
        return transaction

    def send_money(
        self,
        from_user_id: str,
        to_user_id: str,
        amount: float,
        from_account_type: str = "checking",
        to_account_type: str = "checking",
        description: str = None
    ) -> Transaction:
        """Send money to another user."""
        if from_user_id == to_user_id:
            raise HTTPException(
                status_code=400,
                detail="Cannot send money to yourself. Use transfer instead."
            )

        if from_account_type == "treasure_chest" or to_account_type == "treasure_chest":
            raise HTTPException(
                status_code=400,
                detail="Cannot send/receive gold bars. Use checking or savings."
            )

        # Validate users exist
        from_user = self.db.query(User).filter(User.id == from_user_id).first()
        to_user = self.db.query(User).filter(User.id == to_user_id).first()

        if not from_user:
            raise HTTPException(status_code=404, detail="Sender not found")
        if not to_user:
            raise HTTPException(status_code=404, detail="Recipient not found")

        from_account = self.account_service.get_account_by_type(from_user_id, from_account_type)
        to_account = self.account_service.get_account_by_type(to_user_id, to_account_type)

        if from_account.balance < amount:
            raise HTTPException(status_code=400, detail="Insufficient funds")

        from_account.balance -= amount
        to_account.balance += amount

        transaction = Transaction(
            from_account_id=from_account.id,
            to_account_id=to_account.id,
            amount=amount,
            type="transfer",
            description=description or f"Transfer from {from_user.name} to {to_user.name}"
        )
        self.db.add(transaction)
        self.db.commit()
        self.db.refresh(transaction)
        return transaction

    def get_transaction_history(self, user_id: str, limit: int = 50) -> list[Transaction]:
        """Get transaction history for a user."""
        accounts = self.db.query(Account).filter(Account.user_id == user_id).all()
        account_ids = [a.id for a in accounts]

        if not account_ids:
            return []

        transactions = self.db.query(Transaction).filter(
            (Transaction.from_account_id.in_(account_ids)) |
            (Transaction.to_account_id.in_(account_ids))
        ).order_by(Transaction.created_at.desc()).limit(limit).all()

        return transactions

    def get_account_transactions(self, account_id: int, limit: int = 50) -> list[Transaction]:
        """Get transactions for a specific account."""
        transactions = self.db.query(Transaction).filter(
            (Transaction.from_account_id == account_id) |
            (Transaction.to_account_id == account_id)
        ).order_by(Transaction.created_at.desc()).limit(limit).all()

        return transactions
